x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java

```diff
@@ -423,6 +423,8 @@ else if (value instanceof IntervalDayTime
                     else {
                         throw new SqlIllegalArgumentException("Cannot GROUP BY function {}", exp);
                     }
+                } else if (exp instanceof Literal) {
+                    key = new GroupByValue(aggId, QueryTranslator.nameOf(exp));
```

Comment:
I'm a bit surprised that this actually works but apparently the terms aggregation just ignores fields that do not exist and creates a `null` bucket. I'm not sure though wether this might not introduce other bugs. E.g. if `QueryTranslator.nameOf(exp)` evaluates to an actual field name. Could it be an option to have a `GroupByConstant` for this case?

Reply:
`GroupByConstant` was my first idea as well (I also implemented it with the exact same name at first), but then it ended up being the exact copy of GroupByValue. 

The fact that also scalar functions use GroupByValue (passing the function as a script, see the other constructor) is another pointer that convinced me to use it.

Unless there is some practical differentiation, probably it can be avoided (anyway, I don't have a strong opinion on that).


About the clash with field names, I reasoned a bit about that and did some tests.


On a `person (name, surname)` schema, the following are possible clashes

```
SELECT 1 as name FROM person GROUP BY name; // the alias is the same as a field name 
```

That is interpreted correctly

```
SELECT 1 FROM person GROUP BY 'name'; // the value of the literal is the same as a field name
```
Here the field name in the terms is `'name'`, so there is no clash with the field `name`.

Still, I just realised that you could define a field named `'name'`(!!!). In this case, with the same query you would get an error:
`Text fields are not optimised for operations that require per-document field data like aggregations and sorting, so these operations are disabled by default. Please use a keyword field instead. Alternatively, set fielddata=true on ['name'] in order to load field data by uninverting the inverted index. Note that this can use significant memory`.

So I think I'll have to specialise the behaviour

Reply:
Other cases that come to my mind are fields named `null`, `true` or `false` (I think ES does also not stop you from doing this).